src/V3Ast.h

```diff
@@ -2108,71 +2109,115 @@ class AstNode VL_NOT_FINAL {
     // handle a single (or a few) node types, as it's easier to write, but more importantly, the
     // dispatch to the operation function in 'foreach' should be completely predictable by branch
     // target caches in modern CPUs, while it is basically unpredictable for VNVisitor.
-    template <typename T_Node>
-    void foreach (std::function<void(T_Node*)> f) {
-        static_assert(checkTypeParameter<T_Node>(), "Invalid type parameter 'T_Node'");
+    template <typename Func>
+    void foreach (Func&& f) {
+        using traits = vlstd::function_traits<decltype(f)>;
+        using T = typename traits::template arg<0>::type;
+        using T_Node = typename std::remove_pointer<T>::type;
+        static_assert(checkTypeParameter<typename std::remove_const<T_Node>::type>(),
+                      "Invalid type parameter 'T_Node'");
+        static_assert(vlstd::is_invocable<Func, T>::value, "Func must be invocable");
```

Comment:
I would like a single static_assert here of the form:
`static_assert(<condition>, "'Callable' must have a signature 'void(T_Node *)')`
In particular: signature matches old version, and the static_assert message says what is expected, otherwise it is hard to derive.

Same applies elsewhere.